### PR TITLE
add assume-role credential provider

### LINKED ISSUE #817: Minio AssumeRole using SDK
Is there support for the AssumeRole funtion from either the Minio Java SDK or the AWS Java SDK?

I found no documentation for this in the Minio SDK

I have attempted to use the AWS Java SDK to make an assumeRole call to a Minio sever.
It consistently fails with:
com.amazonaws.services.securitytoken.model.AWSSecurityTokenServiceException: Generating temporary credentials not allowed for this request. (Service: AWSSecurityTokenService; Status Code: 403; Error Code: AccessDenied; Request ID: ########

I am hoping that there SDK support and perhaps an example is available. Thanks.

## Context
We are attempting to get credentials in a java app that a python client can subsequently use to put and get large files.

### PR DESCRIPTION
**Pull request by balamurugana**
Fixes #817

### CODE REVIEW

**Comment by aalex21** [comment]
Can we support the "ExternalId" parameter as part of AssumeRole Query Params ?
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html

This will limit one minio client to be associated with one externalId !!

**Comment by balamurugana** [comment]
> Can we support the "ExternalId" parameter as part of AssumeRole Query Params ?
> https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html
> 
> This will limit one minio client to be associated with one externalId !!

Sure. Added now.

File: api/src/main/java/io/minio/credentials/AssumeRoleProvider.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+public class AssumeRoleProvider implements Provider {
+  public static final int DEFAULT_DURATION_SECONDS = (int) TimeUnit.HOURS.toSeconds(1);
+  private final HttpUrl stsEndpoint;
+  private final String accessKey;
+  private final String secretKey;
+  private final String region;
+  private final OkHttpClient httpClient;
+  private final String contentSha256;
+  private final Request request;
+  private Credentials credentials;
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/AssumeRoleProvider.java:57]: The fields `stsEndpoint`, `httpClient` and `credentials` seem to be common with `WebIdentityClientGrantsProvider`. We can consider adding an abstract base class that has a constructor that takes these arguments and call it from the constructors of both `WebIdentityClientGrantsProvider` and `AssumeRoleProvider`

File: api/src/main/java/io/minio/credentials/AssumeRoleProvider.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    String data = urlBuilder.build().encodedQuery();
+    this.contentSha256 = Digest.sha256Hash(data);
+    this.request =
+        new Request.Builder()
+            .url(this.stsEndpoint)
+            .method(
+                "POST",
+                RequestBody.create(MediaType.parse("application/x-www-form-urlencoded"), data))
+            .build();
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/AssumeRoleProvider.java:122]: The request building logic is inside the `fetch()` method in the `WebIdentityClientGrantsProvider` class. Can we try to have a consistent pattern in both? I see a small potential for an abstract base class that provides the `fetch()` logic as well, forcing the concrete sub-classes to build the request, which seems to be the main difference.

**Review by anjalshireesh — DISMISSED** [review]
LGTM

**Review by sinhaashish — APPROVED** [review]
Tested & LGTM

**Review by anjalshireesh — APPROVED** [review]
LGTM

**Pull request merged**